src/lld/moviebooking/class_diagram.puml

```diff
@@ -1,3 +1,323 @@
+        - id: long
+        - name: String
+    }
+    class User extends Identifier{
+        - email: String
+        - phoneNo: String
+    }
+    class Movie extends Identifier{
+        - crew: List<Crew>
+        - supportedLanguages: List<LanguageCode>
+    }
+    class Crew extends Identifier{
+        - crewType: CrewType
+    }
+    enum CrewType{
+        Cameraman
+        Designer
+        Hero
+        Heroine
+        Director
+        Singer
+    }
+    'Todo: Preferred hierarchical approach, should discuss and final best entity design for city state, this takes four level joins
+    class District extends Identifier{
+        cityId: long
+    }
+    class City extends Identifier{
+        stateId: long
+    }
+    class State extends Identifier{
+        countryId: Country
+    }
+    class Country extends Identifier{
+    }
+
+    class Theatre extends Identifier {
+        - theatreGroupId: long
+        - address: Address
+        - screens: List<Screen>
+        - theatreStatus: TheatreStatus
+    }
+    class TheatreGroup extends Identifier{
+        - founder: String
+        - foundedDate: LocalDateTime
+    }
+    class Address{
+        - id: long
+        - addressLine1: String
+        - addressLine2: String
+        - districtId: districtId
+        - gpsLocation: GpsLocation
+        - pinCode: String
+    }
+    class Screen extends Identifier{
+        - seats: List<Seat>
+        - screenStatus: ScreenStatus
+    }
+    enum TheatreStatus{
+        Available
+        NotAvailable
+    }
+    enum ScreenStatus{
+        Available
+        NotAvailable
+    }
+    'Todo: Seat arrangement belongs to the screen style, user can customise it'
+    class Seat extends Identifier{
+        - rowNo: String
+        - colNo: String
+        - seatStatus: SeatStatus
+    }
+    enum SeatStatus{
+        Available,
+        NotAvailable,
+        Locked
+    }
+    enum SeatCategory{
+        Normal
+        Premium
+    }
+    class Show extends Identifier{
+        - showTime: LocalDateTime
+        - screenId: long
+    }
+}
+
+package com.movie_booking.entity.mapping{
+    class Booking {
+        - bookingId: long
+        - showId: long
+        - seatId: long
+        - userId: long
+        - movieId: long
+        - theatreId: long
+        - screenId: long
+        - bookingTime: LocalDateTime
+        - paymentId: long
+        - bookingStatus: BookingStatus
+    }
+    enum BookingStatus{
+        Booked,
+        Cancelled
+    }
+    class SeatDTO{
+        seat: Seat,
+        showId: long
+    }
+    class TheatreShowDTO{
+        theatre: Theatre,
+        shows: List<Show>
+    }
+    class MovieDTO{
+        movie: Movie
+        supportedLanguagesInCity: List<City>
+        cityWiseDetails:
+    }
+    class cityWiseDetails{
+        cityId: long
+        releaseTime: LocalDateTime
+    }
+}
+package com.movie_booking.dao.main{
+    interface ReadDao<T>{
+        + findById(id: long): T
```

Comment:
should return optional<T> or need to handle exception of entityNotFound

Reply:
** I sought recommendation from chatgpt it says optional is most viable option** 
Returning an Optional or throwing an exception both have their uses and it ultimately depends on the context in which they are used. Here are some factors to consider when deciding between the two:

Handling null values: If the method you are implementing can legitimately return null, then returning an Optional makes sense. The caller can then decide what to do with the null value. If null is not a valid return value, then throwing an exception is a better approach.

Checked vs unchecked exceptions: If the error that you want to communicate to the caller is recoverable, then a checked exception is a good choice. The caller is forced to handle the exception and take appropriate action. If the error is not recoverable, then an unchecked exception like RuntimeException is appropriate.

Performance: Throwing an exception can be expensive in terms of performance, so if you are concerned about performance, returning an Optional may be a better choice.

API design: If you are designing an API for others to use, returning an Optional can make the API more self-documenting. If you are using an API, throwing an exception can be easier to handle since you can catch the exception and take appropriate action.

In general, if the method can legitimately return null and the caller can handle the null value, returning an Optional is a good choice. If null is not a valid return value or if you need to communicate an error that is not recoverable, then throwing an exception is a better approach.

Reply:
so changing optional to all non-list return values